### PR TITLE
=str #19193 remove buffer from InputStreamSource

### LINKED ISSUE #19193: Source.inputStream must emit bytes eagerly without batching
The default behaviour of `Source.inputStream()` is to block until the internal buffer is filled up.

Since we won't be able to tell the length of the data in advance, the sensible default behaviour is to send data as soon as it's available from the Source, instead of waiting until the buffer is filled.

This can be achieved at the moment with the following snippet - (thanks @2m):

```
Source.inputStream(() => input).withAttributes(Attributes.inputBuffer(1, 1))
```

We should have the `Attributes.inputBuffer(1, 1)` as the default behaviour.

### PR DESCRIPTION
**Pull request by 2m**
Fixes #19193

### CODE REVIEW

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala [hunk 1]
```diff
@@ -96,23 +59,20 @@ private[akka] class InputStreamPublisher(is: InputStream, bytesReadPromise: Prom
...
+        onCompleteThenStop()
 
       case _ ⇒
         readBytesTotal += readBytes
-        if (readBytes == chunkSize) availableChunks :+= ByteString1C(arr)
-        else availableChunks :+= ByteString1C(arr).take(readBytes)
 
-      // valid read, continue
+        // emit immediately, as this is the only chance to do it before we might block again
+        onNext(ByteString1C(arr).take(readBytes))
```
Comment (rkuhn) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala]: Isn’t it cheaper to keep using the same array of chunkSize for reading and copying into an appropriately-sized one before emitting? Especially for workloads where this change matters readBytes will frequently be significantly less than chunkSize, wasting a lot of space.

**Comment by rkuhn** [comment]
Thanks @2m !

/cc @fsat

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala (hunk 1 above)
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala]: Using the same array might be also more cache efficient. Not that it matters for IO though...

**Comment by drewhk** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala (hunk 1 above)
Comment (2m) [inline akka-stream/src/main/scala/akka/stream/impl/io/InputStreamPublisher.scala]: Let's see. With the shared array the worst case scenario is if the `chunkSize` is such that `is.read(..)` almost always reads fully, then the space wasted is the size of the shared array.

With no shared array wasted space is for every read of size `chunkSize - readBytes`. Which could be small but also does not have a limit.

It really looks like shared array is less wasteful.

**Comment by 2m** [comment]
Changed impl to have a shared buffer.

@viktorklang I remember we had a discussion against this before, but in the current light it looks like shared buffer approach produces at least limited memory waste.

**Comment by drewhk** [comment]
The previous discussion was about shared array vs ByteString1C (which does not do a copy of the array).

**Pull request merged**